core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala

```diff
@@ -163,32 +157,38 @@ class LoadBalancerService(
                 processCompletion(Left(activationId), transid, forced = true)
             }
 
-            ActivationEntry(activationId, namespaceId, invokerName, Instant.now(Clock.systemUTC()), Promise[Either[ActivationId, WhiskActivation]]())
-        })
+            ActivationEntry(activationId, namespaceId, invokerName, Instant.now(Clock.systemUTC()),
+                Promise[Either[ActivationId, WhiskActivation]]())
+
+    }
+
+    /**
+     * Creates an activation entry and insert into various maps.
+     */
+    private def setupActivation(action: ExecutableWhiskAction, activationId: ActivationId, namespaceId: UUID, invokerName: String, transid: TransactionId): ActivationEntry = {
+        // either create a new promise or reuse a previous one for this activation if it exists
+        val entry = loadBalancerData.getActivationByIdOrUpdate(activationId, () => setTimeout(action, activationId,
+            namespaceId, invokerName, transid))
 
         // add the entry to our maps, for bookkeeping
-        activationByInvoker.getOrElseUpdate(invokerName, new TrieSet[ActivationEntry]).put(entry, {})
-        activationByNamespaceId.getOrElseUpdate(namespaceId, new TrieSet[ActivationEntry]).put(entry, {})
+        loadBalancerData.putActivation(entry)
         entry
     }
 
     /**
-     * When invoker health detects a new invoker has come up, this callback is called.
-     */
+      * When invoker health detects a new invoker has come up, this callback is called.
+      */
     private def clearInvokerState(invokerName: String) = {
-        val actSet = activationByInvoker.getOrElseUpdate(invokerName, new TrieSet[ActivationEntry])
+        val actSet = loadBalancerData.activationsByInvoker(invokerName)
         actSet.keySet map {
             case actEntry @ ActivationEntry(activationId, namespaceId, invokerIndex, _, promise) =>
                 promise.tryFailure(new LoadBalancerException(s"Invoker $invokerIndex restarted"))
-                activationById.remove(activationId)
-                activationByNamespaceId.get(namespaceId) map { _.remove(actEntry) }
+                loadBalancerData.removeActivationById(activationId)
+                loadBalancerData.removeActivation(actEntry)
```

Comment:
I think the duplication is valid here. We need to remove the activation from the first map first:
```
        loadBalancerData.removeActivationById(aid) match {
        case Some(entry @ ActivationEntry(_, namespaceId, invokerIndex, _, p)) =>
            logging.info(this, s"${if (!forced) "received" else "forced"} active ack for '$aid'")(tid)
            loadBalancerData.removeActivation(entry)
```